src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp

```diff
@@ -14,320 +14,173 @@
+        const String & queue_base_,
         Poco::Logger * log_,
         char row_delimiter_,
-        bool bind_by_id_,
+        bool hash_exchange_,
         size_t num_queues_,
-        const String & exchange_type_,
         const String & local_exchange_,
+        const String & deadletter_exchange_,
         const std::atomic<bool> & stopped_)
         : ReadBuffer(nullptr, 0)
         , consumer_channel(std::move(consumer_channel_))
+        , setup_channel(setup_channel_)
         , event_handler(event_handler_)
         , exchange_name(exchange_name_)
+        , exchange_type(exchange_type_)
         , routing_keys(routing_keys_)
         , channel_id(channel_id_)
-        , bind_by_id(bind_by_id_)
+        , queue_base(queue_base_)
+        , hash_exchange(hash_exchange_)
         , num_queues(num_queues_)
-        , exchange_type(exchange_type_)
-        , local_exchange(local_exchange_)
-        , local_default_exchange(local_exchange + "_" + ExchangeType::DIRECT)
-        , local_hash_exchange(local_exchange + "_" + ExchangeType::HASH)
         , log(log_)
         , row_delimiter(row_delimiter_)
         , stopped(stopped_)
-        , messages(QUEUE_SIZE * num_queues)
+        , local_exchange(local_exchange_)
+        , deadletter_exchange(deadletter_exchange_)
+        , received(QUEUE_SIZE * num_queues)
 {
-    exchange_type_set = exchange_type != ExchangeType::DEFAULT;
-
-    /* One queue per consumer can handle up to 50000 messages. More queues per consumer can be added.
-     * By default there is one queue per consumer.
-     */
     for (size_t queue_id = 0; queue_id < num_queues; ++queue_id)
-    {
-        /// Queue bingings must be declared before any publishing => it must be done here and not in readPrefix()
         initQueueBindings(queue_id);
-    }
 }
 
 
 ReadBufferFromRabbitMQConsumer::~ReadBufferFromRabbitMQConsumer()
 {
     consumer_channel->close();
-
-    messages.clear();
+    received.clear();
     BufferBase::set(nullptr, 0, 0);
 }
 
 
-void ReadBufferFromRabbitMQConsumer::initExchange()
+void ReadBufferFromRabbitMQConsumer::initQueueBindings(const size_t queue_id)
 {
-    /* This direct-exchange is used for default implemenation and for INSERT query (so it is always declared). If exchange_type
-     * is not set, then there are only two exchanges - external, defined by the client, and local, unique for each table (default).
-     * This strict division to external and local exchanges is needed to avoid too much complexity with defining exchange_name
-     * for INSERT query producer and, in general, it is better to distinguish them into separate ones.
-     */
-    consumer_channel->declareExchange(local_default_exchange, AMQP::direct).onError([&](const char * message)
-    {
-        local_exchange_declared = false;
-        LOG_ERROR(log, "Failed to declare local direct-exchange. Reason: {}", message);
-    });
+    std::atomic<bool> bindings_created = false, bindings_error = false;
```

Comment:
Why they have to be atomic?

Reply:
because otherwise later in the loop you'll catch `error: this loop is infinite; none of its condition variables (bindings_created, bindings_error) are updated in the loop body`